src/quill-editor.component.ts

```diff
@@ -261,9 +261,12 @@ export class QuillEditorComponent
         }
 
         this.zone.run(() => {
-          this.onModelChange(
-            this.valueGetter(this.quillEditor, this.editorElem)
-          );
+
+          if (source === 'user') {
```

Comment:
But i have to call it for every changes... to keep the model in sync with the displayed view value. See my tests.

even wheb someone is using the quill api to change the content, it should be represented in the model

Reply:
It is too much to answer at once :)

Reply:
> This not solves the problem: "when using ngModel the model is marked as dirty"
I believe it does. I guess I will have to setup a demo page.

Reply:
If your idea was to emulate user interaction with quill, you could at least call `setContents`, passing `source='user'`. But as you use a real browser to launch unit tests, DOM event emulation seems a better choice (because it emulates user actions as closely as possible with the current test setup).

Reply:
> And this event should be triggered after the intital value is set.

As you mentioned in https://github.com/KillerCodeMonkey/ngx-quill/issues/121, it is not clear what value to consider initial. The Angular rules are clear and do not have anything to do with 'initial' and 'non-initial' values.

Reply:
Summary:

> But i have to call it for every change... to keep the model in sync with the displayed view value. See my tests.

Disagree. My point is that in your tests you do not emulate user interaction.

> even when someone is using the quill api to change the content, it should be represented in the model

My point is that calling the quill api directly should **not** be considered a legitimate use of an Angular component.

> WriteValue keeps the view in sync with the model
> modelChangedFn keeps the model in sync with the view

Agree.

> And this event should be triggered after the intital value is set.

Agree, we do not discern initial and non-initial values.

> This not solves the problem: "when using ngModel the model is marked as dirty"

Disagree. Going to prove it with a demo page :)

Reply:
There is no need for the demo ;). i can imagine it know, but you could simply extend a test case, to check if the model is pristine.

Reply:
@earshinov ahh the problem is, that all formats and toolbar interactions are not triggered as "user" changes. Maybe this was my main intention to not filtering only user interactions.

Because those changes should be represented on the model ;)

Reply:
@KillerCodeMonkey , ouch, yes, the issue with toolbar buttons etc. is crucial. I guess I should spend more time on my changes and cover more scenarios.

By the way, do you happen to have an official demo app? Like [ng-select](https://github.com/ng-select/ng-select/tree/master/demo) or [angular-archwizard](https://github.com/ng-select/ng-select/tree/master/demo)?

Reply:
@earshinov i tested it again. it should worl with the toolbar buttons? or am i wrong

Reply:
> There is no need for the demo ;). i can imagine it know, but you could simply extend a test case, to check if the model is pristine.

I've made a demo page: https://stackblitz.com/edit/ngx-quill-dirtiness-demo
(had to include ngx-quill source inline to apply my modifications). From this you can observe that with my modifications the template-driven form remains pristine when the page is opened.

Reply:
> And does a toolbar Change triggers 'user' Change?

Yes, as far as I can see, it actually does. I added logging to the text-change event handler, and what I see after clicking 'Bold' is that

- the form is marked dirty;
- there is "Received text-change; source=user" message in console logs.

Reply:
> @earshinov i tested it again. it should worl with the toolbar buttons? or am i wrong

Yes, it works. In the gif below, the form is correctly marked dirty when the user presses toolbar buttons. Quill correctly sends `source='user'` when a change was initiated by the user, even if he/she triggers the change by pressing a toolbar button rather than by editing text.

![applying bold makes form dirty](https://user-images.githubusercontent.com/52021/49397665-4e1f0b80-f74d-11e8-9b9f-75e33be631ab.gif)

(Note the `source === 'user'` check on the left side)